core/src/providers/openai.rs

```diff
@@ -74,13 +75,215 @@ pub struct Completion {
+                        name: s.to_string(),
+                    },
+                };
+                Ok(OpenAIToolChoice::OpenAISpecificFunction(function))
+            }
+        }
+    }
+}
+
+// Outputs types.
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct OpenAIToolCallFunction {
+    name: String,
+    arguments: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct OpenAIToolCall {
+    r#type: OpenAIToolType,
+    pub function: OpenAIToolCallFunction,
+}
+
+impl TryFrom<&ChatFunctionCall> for OpenAIToolCall {
+    type Error = anyhow::Error;
+
+    fn try_from(cf: &ChatFunctionCall) -> Result<Self, Self::Error> {
+        Ok(OpenAIToolCall {
+            r#type: OpenAIToolType::Function,
+            function: OpenAIToolCallFunction {
+                name: cf.name.clone(),
+                arguments: cf.arguments.clone(),
+            },
+        })
+    }
+}
+
+// This code performs a type conversion with information loss when converting to ChatFunctionCall.
+// It only supports one tool call, so it takes the first one from the vector of OpenAIToolCall,
+// hence potentially discarding other tool calls.
+impl TryFrom<&Vec<OpenAIToolCall>> for ChatFunctionCall {
+    type Error = anyhow::Error;
+
+    fn try_from(tcs: &Vec<OpenAIToolCall>) -> Result<Self, Self::Error> {
+        // We only take the first tool call suggestion.
+        tcs.first()
+            .map(|tc| ChatFunctionCall {
+                name: tc.function.name.clone(),
+                arguments: tc.function.arguments.clone(),
+            })
+            .ok_or_else(|| anyhow::anyhow!("No OpenAIToolCall found"))
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct OpenAIChatMessage {
+    pub role: ChatMessageRole,
+    pub content: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub tool_calls: Option<Vec<OpenAIToolCall>>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct OpenAIChatChoice {
+    pub message: OpenAIChatMessage,
+    pub index: usize,
+    pub finish_reason: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct OpenAIChatCompletion {
+    pub id: String,
+    pub object: String,
+    pub created: u64,
+    pub choices: Vec<OpenAIChatChoice>,
+    // Usage is not returned by the Chat/Completion endpoint when streamed.
+    // pub usage: Usage,
+}
+
+impl TryFrom<&OpenAIChatMessage> for ChatMessage {
+    type Error = anyhow::Error;
+
+    fn try_from(cm: &OpenAIChatMessage) -> Result<Self, Self::Error> {
+        let role = ChatMessageRole::from(cm.role.clone());
+        let content = match cm.content.as_ref() {
+            Some(c) => Some(c.clone()),
+            None => None,
+        };
+
+        let function_call = match cm.tool_calls.as_ref() {
```

Comment:
I think this would be a much better place to do the information loss thing.

Reply:
Moved here.